Allow configuring container process UID

orangejulius opened the pull request:
By default, all Pelias images now have a user called `pelias` with UID 1000. However, running as a different UID may be best for different people, depending on their system setup.

By setting the default in `.env` and using it in `docker-compose.yml`, we can allow people running Pelias to use whatever non-root user is best for them.

Connects https://github.com/pelias/baseimage/pull/2

missinglink commented:
Looks good, we should probably roll it out across the other projects too.

missinglink commented:
@orangejulius I had a couple more thoughts about this:

- adding a variable in `.env` will cause a merge conflict for everyone when pulling
- it [might be possible](https://github.com/docker/compose/issues/2441) to avoid this by using a default value in the `variable substitution` as per bash
- as per my comment above, we should standardize this across all projects
- we should add documentation in the `Configure Environment` section of the README which explains whats the options are, provides links to Docker documentation and explains the whole UID mess.

missinglink commented:
another point against `.env` is the question of what happens if a user is using an old `.env` file which doesn't contain the `DOCKER_USER` variable.

in that case does it fall back to `1000` or does it fail? it it falls back then we can probably omit it and simply document it as optional.

If omission causes failure then we should consider ways to mitigate the failure as it would break backwards compatibility.

orangejulius commented:
These are all good concerns. I consider this code still WIP (now reflected in the title). So we don't want to roll it out to all projects until we've decided what's best.

Some comments:
>adding a variable in .env will cause a merge conflict for everyone when pulling

Yes, ideally we figure out a better way to handle that, as even now I have to constantly stash/unstash my changes to the `DATA_DIR` when switching branches or projects

>it might be possible to avoid this by using a default value in the variable substitution as per bash

Not a bad idea. I'd like to keep as much logic out of `docker-compose.yml` as possible. It's getting large and bogged down with lots of "details", but this doesn't add too much.

>another point against .env is the question of what happens if a user is using an old .env file which doesn't contain the DOCKER_USER variable.

If the variable is omitted completely, it does indeed fall back to the `pelias` user (UID 1000) as defined in the docker images. However, the first line of output from `pelias compose up` is `WARNING: The DOCKER_USER variable is not set. Defaulting to a blank string.` so I don't think that's a good solution.

I see a couple options, none of them perfect:
- use this code essentially as is. It mostly follows our existing patterns and doesn't introduce any new concepts or problems, but amplifies existing pain points
- handle managing a `-u $DOCKER_USER` or equivalent parameter into all docker-compose up/run commands in the `pelias` utility. Then we can add as much logic as we desire, ensure there are no unnecessarily frightening warnings, and handle defaults in a clean way. I don't actually see a flag to set the user id when running `docker-compose up`, but there must be a way.

orangejulius commented:
Hey @fpurcell,
Glad to hear it's working well for you overall.

- looking at the [docs for the user parameter in docker](https://docs.docker.com/engine/reference/run/#user), it appears you can specify a group id as well. Try something like `10000:10000` in the `DOCKER_USER` env var and see if it helps.
- I'm not sure what's happening with the blacklist directory. It could be something related to docker volumes (directories created by mounting a volume are usually owned by root), or maybe one of our containers isn't using a non-root user yet.
- We happen to have a fix coming for the NPM issue. Check out the PRs attached to https://github.com/pelias/pelias/issues/745. We will be merging them soon.

missinglink commented:
We discussed this on a call and I forgot to write it up in here, upon reflection I am :+1: for having it explicitly called out in `.env` and `docker-compose.yml`, those files act as a form of documentation and it makes it easy for new users to understand which settings are available to modify.

orangejulius commented:
This PR is now out of `WIP` as I've updated all projects and added some documentation on the change. 

I'd love another set of eyes on the documentation section before merging.
